rir/src/compiler/translations/pir_2_rir.cpp

```diff
@@ -0,0 +1,601 @@
+                                             // the closure?
+
+                std::vector<FunIdxT> callArgs;
+                call->eachCallArg([&](Value* arg) {
+                    auto mkarg = MkArg::Cast(arg);
+                    callArgs.push_back(promises[mkarg->prom]);
+                });
+
+                cs.insertCall(Opcode::call_, callArgs, {},
+                              Pool::get(call->srcIdx));
+                store(it, call);
+                break;
+            }
+            case Tag::StaticCall: {
+                assert(false && "not yet implemented.");
+                break;
+            }
+            case Tag::StaticEagerCall: {
+                assert(false && "not yet implemented.");
+                break;
+            }
+            case Tag::CallBuiltin: {
+                auto blt = CallBuiltin::Cast(instr);
+                // TODO: is it needed to set the environment?
+                loadEnv(it, blt->env());
+                cs << BC::setEnv();
+                blt->eachCallArg([&](Value* arg) { load(it, arg); });
+                cs.insertStackCall(Opcode::static_call_stack_, blt->nCallArgs(),
+                                   {}, Pool::get(blt->srcIdx), blt->blt);
+                store(it, blt);
+                break;
+            }
+            case Tag::CallSafeBuiltin: {
+                assert(false && "not yet implemented.");
+                break;
+            }
+            case Tag::MkEnv: {
+                auto mkenv = MkEnv::Cast(instr);
+                loadEnv(it, mkenv->parent());
+                cs << BC::makeEnv() << BC::dup() << BC::setEnv();
+                // bind all args
+                mkenv->eachLocalVar([&](SEXP name, Value* val) {
+                    load(it, val);
+                    cs << BC::stvar(name);
+                });
+                store(it, mkenv);
+                break;
+            }
+            case Tag::Phi: {
+                auto phi = Phi::Cast(instr);
+                load(it, phi->arg(0).val());
+                store(it, phi);
+                break;
+            }
+            case Tag::Deopt: {
+                assert(false && "not yet implemented.");
+                break;
+            }
+            // values, not instructions
+            case Tag::Missing:
+            case Tag::Env:
+            case Tag::Nil:
+                break;
+            // dummy sentinel enum item
+            case Tag::_UNUSED_:
+                break;
+            // no default to get compiler warnings for missing...
+            }
+        }
+
+        // this BB has exactly one successor, next0
+        // jump through empty blocks
+        auto next = bb->next0;
+        while (next->isEmpty())
+            next = next->next0;
+        cs << BC::br(bbLabels[next]);
+    });
+
+    return alloc.slots();
+}
+
+void Pir2Rir::removePhis(Code* code) {
+
+    // For each Phi, insert copies
```

Comment:
ah, yeah I was also struggling with the name here, but then forgot, after I understood what it does :)  Boissinot etal calls this "converting from SSA to CSSA (conventional ssa)"

Reply:
btw. (but that's for later), the copies should be parallel, to avoid introducing artificial interference.

I think we should have the following:

every BB has an "out-copy" and an "in-phi" table. Then the conversion would look like this:

```
BB0:
  a0 = 0
  b0 = 0

BB1:
  goto BB3

BB2:
  a2 = 2
  b2 = 2
  goto BB3

BB3:
  x = phi(a0, b2)
  y = phi(a2, b0)
```

in this case we would say:

```
BB1-out:
    a1o, b1o = a0, b0

BB2-out:
    a2o, b20 = a2, b2

BB3-in:
    x, y = a1o | b2o , a2o | b1o
```

Reply:
then, basically the removePhis pass would (instead of inserting copy instructions) populate the in-/out- tables